File: Language/C#/Book/tmp/LINQ实战/Chapter06to08.CSharp/Chapter06to08/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using LinqInAction.Chapter06to08.Common;
using LinqInAction.Chapter06to08.Common.CodeSamples;
using LinqInAction.Chapter06to08.Common.SampleHarness;

namespace Samples.CSharp
{
    public partial class MainForm : Form
    {

        public MainForm()
        {
            InitializeComponent();
        }

        private List<SampleClass> GetSamples()
        {
            List<SampleClass> sampleList = new List<SampleClass>();
            sampleList.Add(new Ch6Samples());
            sampleList.Add(new Ch7Samples());
            sampleList.Add(new Ch8Samples());
            return sampleList;
        }

        private void Form_Load(object sender, EventArgs e)
        {
            List<SampleClass> sampleList = GetSamples();
            foreach (SampleClass sampleClass in sampleList)
            {
                TreeNode classNode = this.SampleTreeview.Nodes.Add(sampleClass.Title);
                classNode.Tag = sampleClass;
                foreach (SampleItem item in sampleClass)
                {
                    TreeNode itemNode = classNode.Nodes.Add(item.Description);
                    itemNode.Tag = item;
                }
            }
        }

        SampleItem currentSample;
        SampleClass currentClass;

        private void SampleTreeview_AfterSelect(object sender, TreeViewEventArgs e)
        {
            TreeNode selectedNode = SampleTreeview.SelectedNode;
            if (selectedNode.Parent != null)
            {
                currentSample = (SampleItem)selectedNode.Tag;
                if (currentSample != null)
                {
                    currentClass = (SampleClass)selectedNode.Parent.Tag;
                    this.ChapterLabel.Text = currentSample.Chapter.ToString();
                    this.ListingLabel.Text = currentSample.ListingNumber.ToString();
                    this.DescriptionLabel.Text = currentSample.Description;
                    this.btnRun.Enabled = true;
                }
                else
                {
                    this.ChapterLabel.Text = "";
                    this.ListingLabel.Text = "";
                    this.DescriptionLabel.Text = "";
                    this.btnRun.Enabled = false;
                }
            }
        }

        private void btnRun_Click(object sender, EventArgs e)
        {
            OutputTextBos.Text = "";
            if (currentSample != null)
                RunSample(currentClass, currentSample);
        }

        private void btnRunAll_Click(object sender, EventArgs e)
        {
            this.OutputTextBos.Text = "";
            List<SampleClass> sampleList = GetSamples();
            foreach (SampleClass sampleClass in sampleList)
                foreach (SampleItem item in sampleClass)
                    RunSample(sampleClass, item);

        }

        private void RunSample(SampleClass currentClass, SampleItem currentSample)
        {
            if (currentSample != null)
            {
                base.UseWaitCursor = true;
                StreamWriter newOut = currentClass.OutputStreamWriter;
                TextWriter output = Console.Out;
                Console.SetOut(newOut);
                MemoryStream baseStream = (MemoryStream)newOut.BaseStream;
                baseStream.SetLength(0);
                try
                {
                    currentSample.Method.Invoke(currentClass, null);
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.InnerException.ToString());
                }
                newOut.Flush();
                Console.SetOut(output);
                OutputTextBos.Text = OutputTextBos.Text + newOut.Encoding.GetString(baseStream.ToArray());
                base.UseWaitCursor = false;
            }
        }
    }
}

File: Language/C#/Book/LINQ实战/第06章 LINQ to SQL入门/Chapter06to08.VB/Chapter06to08/MainForm.vb
Option Strict On

Imports LinqInAction.Chapter06to08.Common
Imports System.IO

Public Class MainForm
  Private _SampleList As New List(Of SampleClass)
  Private Sub Form_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
    _SampleList.Add(New CodeSamples.Ch6Samples)
    _SampleList.Add(New CodeSamples.Ch7Samples)
    _SampleList.Add(New CodeSamples.Ch8Samples)


    For Each _class In _SampleList
      Dim ClassNode = Me.SampleTreeview.Nodes.Add(_class.Title)
      ClassNode.Tag = _class
      For Each item As SampleItem In _class
        Dim ItemNode = ClassNode.Nodes.Add(item.Description)
        ItemNode.Tag = item
      Next
    Next
  End Sub

  Dim CurrentSample As SampleItem
  Dim CurrentClass As SampleClass

  Private Sub SampleTreeview_AfterSelect(ByVal sender As System.Object, ByVal e As System.Windows.Forms.TreeViewEventArgs) Handles SampleTreeview.AfterSelect
    Dim selectedNode As TreeNode = SampleTreeview.SelectedNode
    CurrentSample = TryCast(selectedNode.Tag, SampleItem)

    If CurrentSample IsNot Nothing Then
      CurrentClass = TryCast(selectedNode.Parent.Tag, SampleClass)
      ChapterLabel.Text = CurrentSample.Chapter.ToString
      ListingLabel.Text = CurrentSample.ListingNumber.ToString
      DescriptionLabel.Text = CurrentSample.Description
      btnRun.Enabled = True
    Else
      ChapterLabel.Text = ""
      ListingLabel.Text = ""
      DescriptionLabel.Text = ""
      btnRun.Enabled = False
    End If
  End Sub

  Private Sub btnRun_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRun.Click
    OutputTextBox.Text = ""
    If Not CurrentSample Is Nothing Then
      RunSample(CurrentClass, CurrentSample)
    End If
  End Sub

  Private Sub Button1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Button1.Click
    OutputTextBox.Text = ""
    For Each sampleClass In _SampleList
      For Each item As SampleItem In sampleClass
        RunSample(sampleClass, item)
      Next
    Next
  End Sub

  Private Sub RunSample(ByVal currentClass As SampleClass, ByVal currentSample As SampleItem)
    If Not currentSample Is Nothing Then
      MyBase.UseWaitCursor = True
      Dim newOut As StreamWriter = CurrentClass.OutputStreamWriter
      Dim out As TextWriter = Console.Out
      Console.SetOut(newOut)
      Dim baseStream As MemoryStream = DirectCast(newOut.BaseStream, MemoryStream)
      baseStream.SetLength(CLng(0))
      Try
        currentSample.Method.Invoke(CurrentClass, Nothing)
      Catch ex As Exception
        Console.WriteLine(ex.InnerException.ToString)
      End Try

      newOut.Flush()
      Console.SetOut(out)
      OutputTextBox.Text = (OutputTextBox.Text & newOut.Encoding.GetString(baseStream.ToArray))
      MyBase.UseWaitCursor = False

    End If
  End Sub
End Class
